Using proxy to omit re-render

bluebill1049 opened the pull request:
- 🌹`proxy` reduces extra re-render when you are not reading the formState

bluebill1049 committed:
using proxy to omit render

bluebill1049 committed:
update handleSubmit code

bluebill1049 committed:
fix automations

bluebill1049 committed:
fix type and automation

bluebill1049 commented on cypress/integration/basic.ts:
```diff
@@ -67,7 +67,7 @@ context('basic form validation', () => {
     cy.get('input[name="minDate"]').should('not.have.value');
     cy.get('input[name="maxDate"]').should('not.have.value');
-    cy.get('#renderCount').contains('47');
+    cy.get('#renderCount').contains('2');
```
Look this!

barrymay commented on src/useForm.ts:
```diff
@@ -75,6 +77,14 @@ export default function useForm<FormValues extends FieldValues = FieldValues>({
   const validateAndUpdateStateRef = useRef<Function>();
   const [, render] = useState();
   const { isOnBlur, isOnSubmit } = useRef(modeChecker(mode)).current;
+  const readFormState = useRef({
```
let's make this ref use useRef<FormState>(...) if that's what the format should be.  Others are obvious booleans and such but that would keep any changes consistent

(this is quite an impressive PR - I've never directly used Proxies, but I've used the amazing state lib immer which leverages them and it's a game changer

barrymay replied:
(if you want help with the TS on this pr, I will ;)

bluebill1049 replied:
yes plz @barrymay if u have time :)

barrymay replied:
I'll get to this tomorrow my time ;)

bluebill1049 replied:
sweeeet~ thanks @barrymay

barrymay replied:
we should definitely please try to ensure all useRef's have types on them, esp. complex types like this.   If not you're looking for accidental misunderstandings like my example below.

bluebill1049 replied:
you are right @barrymay having quite a lot of boolean ref made me too lazy to give types for `ref` which i will do <3

JeromeDeLeon committed:
Improve ReadFormState type

bluebill1049 commented on src/useForm.ts:
```diff
@@ -900,26 +922,38 @@ export default function useForm<FormValues extends FieldValues = FieldValues>({
+      {
+        get: (obj: FormState<FieldValues>, prop: ReadFormState) => {
+          if (!(prop in obj)) {
+            return {};
```
hmm i need to read a bit more on this, i am worried return undefined will break apps

JeromeDeLeon replied:
I see. We're not actually breaking the app. Its just common behavior to see `undefined` if you access value the is not defined the in the `formState`.

JeromeDeLeon replied:
I would also like to provide `set` in the Proxy as precautionary since we provided error handler to `get` for accessing undefined values.

bluebill1049 replied:
> I see. We're not actually breaking the app. Its just common behavior to see `undefined` if you access value the is not defined the in the `formState`.

I meant breaking others' APP if it returns undefined (which it shouldn't).

bluebill1049 replied:
> I would also like to provide `set` in the Proxy as precautionary since we provided error handler to `get` for accessing undefined values.

curious to know what's the reason behind `set`?

JeromeDeLeon replied:
On second thought, we don't really need it. Sort of habit. 😞

bluebill1049 replied:
> On second thought, we don't really need it. Sort of habit. 😞

yea you are right, we probably dont need the check. i will do a double check on it

JeromeDeLeon replied:
I think its fine though. if they accidentally refer `toched` instead of `touched`, they will still got an error because they cannot read index of undefined.

bluebill1049 replied:
> I think its fine though. if they accidentally refer `toched` instead of `touched`, they will still got an error because they cannot read index of undefined.

ya, thanks @JeromeDeLeon i will do some testing tonight.

barrymay replied:
chiming in here, I'm also in favor of it not returning empty obj every time - it should be the undefined or at least the type of the prop

bluebill1049 replied:
I think there will be a problem when you run `JSON.stringify(formState)`, the prop is return `toJSON`. I need to read more on the proxy to find out why it's doing that.

bluebill1049 commented on src/types.ts:
```diff
@@ -138,13 +138,7 @@ export interface FormState<FormValues extends FieldValues = FieldValues> {
-  | 'touched'
-  | 'isSubmitting'
-  | 'isValid';
+export type ReadFormState = keyof FormState;
```
amazing! thank you @JeromeDeLeon

bluebill1049 committed:
Merge branch 'master' into proxy-formState

bluebill1049 commented:
if you guys have time, please help testing as well :) `3.25.1-beta.1` This is going to be a really good feature for RHF. ❤️

bluebill1049 committed:
remove unneeded check

barrymay commented on src/useForm.ts:
```diff
@@ -898,26 +918,38 @@ export default function useForm<FormValues extends FieldValues = FieldValues>({
-              : !isEmptyObject(fieldsRef.current),
-          }),
-    },
+          readFormState.current[prop] = true;
```
hi @bluebill1049  - this line is problematic (found this while typing - thanks TS!).  Not all props in the Proxy are bools (e.g. submitCount, touched) - so what exactly is this line doing?

bluebill1049 replied:
hey @barrymay it's a flag, saying 'yes' the user is reading 
`formState: dirty` or `formState: isSubmitted`

bluebill1049 replied:
i will take a look at your branch as well

barrymay replied:
ohh I see - let me revert on branch then and fix my typings then on the ref (what you just said clarifies what that's supposed to be)

bluebill1049 replied:
sweet thanks @barrymay

bluebill1049 commented:
it's a side question: @JeromeDeLeon and @barrymay 

Could I please have your permission to include your guys as the core contributors? You guys contributing to the repo and credit should be given. right now @michaelbirchler and I is on the list.

<img width="786" alt="Screen Shot 2019-10-26 at 9 05 02 am" src="https://user-images.githubusercontent.com/10513364/67607300-c52f1980-f7cf-11e9-920f-07ff479a5dc3.png">

barrymay commented:
Just found an important issue - Proxies are not available in react-native since it's not a web environment.  To do that we're going to need to add a polyfill (the aforementioned immer lib does the same for older browsers and react-native). https://stackoverflow.com/questions/38871456/es6-proxy-not-defined-in-react-native-when-not-in-remote-debug-mode

barrymay commented:
> if you guys have time, please help testing as well :) `3.25.1-beta.1` This is going to be a really good feature for RHF. ❤️

The beta worked fine in my test case

barrymay committed:
Proxy form state typefix (#413)

* Remove unnecessary ts-ignore

* Improve typing of proxy and related ref

* Improve proxy types
- ReadFormState should be the types of the readFormState useRef - the keyof is clear in the prop context

bluebill1049 commented:
thanks @barrymay that was an important find!

bluebill1049 commented:
I will verify it with my react-native test app.

bluebill1049 commented:
@barrymay i have tested in my react-native test app looks like it's working for me. maybe to do with node version?

bluebill1049 commented:
<img width="1674" alt="Screen Shot 2019-10-26 at 12 10 23 pm" src="https://user-images.githubusercontent.com/10513364/67611911-a9d10800-f7e9-11e9-8cb1-d7d3e04800aa.png">

barrymay commented:
the thread I sent you was specifically that it worked in remote debug mode - what happens if you do the same thing in release mode?

barrymay commented:
(little react-native side note, I've been working on getting my own sample app running on Github Actions CI, and I just it working to build my react-native app from scratch on their Mac CI servers!  Feel free to check it out (use branch free-images) and ask questions if you'd like

bluebill1049 commented:
thanks @barrymay I am testing release mode now. your side project looks very interesting and it's using github action. 👀how hard would for us to convert to github actions from cricle ci from your experience?

barrymay commented:
I've done two projects on GitHub Actions (both web and native now) and I've found it super easy to use:

My other project is this:
https://github.com/barrymay/phase-ten-scorer-nextjs
(this is my ongoing personal project using react-hook-form)

The only thing up in the air about GitHub actions is the price.  Right now it's been free for the beta (it's been awesome to use since you can run as many jobs as you like) but I'm sure they'll give more clarity at the GitHub Universe conference in the next few weeks.

bluebill1049 commented:
test with production mode with the beta. looks like it's working as well:

https://expo.io/@bluebill1049/AwesomeProject1

bluebill1049 commented:
> I've done two projects on GitHub Actions (both web and native now) and I've found it super easy to use:
> 
> My other project is this:
> https://github.com/barrymay/phase-ten-scorer-nextjs
> (this is my ongoing personal project using react-hook-form)
> 
> The only thing up in the air about GitHub actions is the price. Right now it's been free for the beta (it's been awesome to use since you can run as many jobs as you like) but I'm sure they'll give more clarity at the GitHub Universe conference in the next few weeks.
> 
> I can't imagine it'll be tough to convert, the primary thing is setting up the yml file under the GitHub folder, and setting up the secrets in your project settings. Once you have that, it's very good and getting better

nice! maybe it's worth to convert it over.

barrymay commented:
sweet (they must have added something new to do that...) I have no problem just going with the proxy as is for now - just keep an eye out if someone who upgrades the project has an issue.  I don't have deep enough knowledge on react-native to fully judge if this will be an issue.

bluebill1049 commented:
tested on production mode: 

> sweet (they must have added something new to do that...) I have no problem just going with the proxy as is for now - just keep an eye out if someone who upgrades the project has an issue. I don't have deep enough knowledge on react-native to fully judge if this will be an issue.

Thanks @barrymay I will keep an eye on it, in fact, I will prepare a PR to back it up, so we can quickly turn around if there is an issue to patch. much appreciated your help as usual. 🌹

barrymay reviewed (approved):
@bluebill1049 other than the raised JSON.stringify concern you raised and were checking - this looks fine to me
